channeldb/invoices.go

```diff
@@ -1237,6 +1252,14 @@ func serializeHtlcs(w io.Writer, htlcs map[CircuitKey]*InvoiceHTLC) error {
 			tlv.MakePrimitiveRecord(mppTotalAmtType, &mppTotalAmt),
 		)
 
+		if htlc.AMP != nil {
+			setIDRecord := tlv.MakeDynamicRecord(
```

Comment:
For the `MPP` records, we flattened the fields. Is there a reason why we'd want the specific wire encoding also in the database? It can be an undesired dependency perhaps.

Reply:
Well it's the same encoding, just now part of a "stream" vs us reading out each of the TLV types one by one (it's one big record as is).

Reply:
It is not consistent with how we store the mpp fields though

Reply:
the payment addr field in MPP is already stored at the invoice level, that's why it's flattened

Reply:
also i wouldn't think of things in `record` as wire encodings specifically, it's just mean to organize common TLV records. we already depend on `record.CustomSet` in channeldb

Reply:
It works, but I think we are creating a needlessly complex stack. First we have bbolt that could already do key/value. We don't use that and instead use a single key with our own tlv encoded data. But we don't use that either and store the amp record as a composite value that is wire-encoded in our own tlv data.

You are right that for MPP there was indeed nothing to decide.

Referencing a type alias in the `record` package is to me something different than using yet another encoding in the database.

If you and @Roasbeef think the current encoding is best, I won't object further.

Reply:
> But we don't use that either and store the amp record as a composite value that is wire-encoded in our own tlv data.

TLV was designed to support composite structs, why is this an issue? Would it be preferable to test all 8 possible combinations of `RootShare`, `SetID` and `ChildIndex` being absent/present during decoding? Or to reduce that to 2 possibilities: AMP present vs AMP not-present?

 Custom records are already stored in their wire format directly. So the only exception to these rules currently is MPP, which was flattened because the payment address is implied from the invoice.

> It works, but I think we are creating a needlessly complex stack. 

The purpose is simply to copy the data in the payload and make it available persistently further down the pipeline. How does implementing _more_ custom serialization code lead to less complexity? More importantly IMO, does it increase or decrease the likelihood of error?

Unless we have a strong reason to diverge from the wire format, e.g. MPP, I don't think we should. I don't foresee this happening too often though since we can always augment the raw record with other fields at the `InvoiceHTLC` level.

Reply:
Another benefit to reusing the wire serialization is that the width of the child_index can be increased w/o a migration, since it uses the truncated integer encoding.

Reply:
>Would it be preferable to test all 8 possible combinations of RootShare, SetID and ChildIndex being absent/present during decoding? 

We don't need to test them all, because we know what we write. If `RootShare` is there, we can assume that everything is there.

>Custom records are already stored in their wire format directly. 

This is an expception. We don't know what is in those records so can only store them as opaque byte arrays.